### ISSUE TITLE
`re-matches` is broken in both directions: its anchoring guard compares a capture-group count to a byte length

### ISSUE THREAD

**Issue by csm** [csm/clojurust#333]
## Problem

`Matcher::next`'s `match_all` guard — the branch that makes `re-matches`
anchoring rather than searching — compares the number of capture groups against
the length of the haystack:

```rust
// crates/cljrs-value/src/regex.rs:70 (main)
if !self.match_all || cap.len() == self.haystack.get().len() {
```

`Captures::len()` is the group count including group 0, so for a pattern with no
groups it is 1, for `(\d+)-(\d+)` it is 3. Comparing that to a byte length is
not a check of anything. `re-matches` is the only caller that passes
`match_all = true` (`builtin_re_matches` → `new_matcher(args, true)`), so the
damage is confined to it — `re-find`, `re-seq` and `re-matcher` all pass `false`
and are correct.

## Observed vs. Clojure

Measured against the interpreter (tiered mode, `regex-full`):

| expression | clojurust | Clojure |
|---|---|---|
| `(re-matches #"\d+" "42")` | `nil` | `"42"` |
| `(re-matches #"\d+" "424")` | `nil` | `"424"` |
| `(re-matches #"a+" "aaa")` | `nil` | `"aaa"` |
| `(re-matches #".*" "hello")` | `nil` | `"hello"` |
| `(re-matches #"(\d+)-(\d+)" "12-345")` | `nil` | `["12-345" "12" "345"]` |
| `(re-matches #"\d+" "4")` | `"4"` | `"4"` — correct by accident (group count 1 == length 1) |
| `(re-matches #"(a)(b)" "abc")` | `["ab" "a" "b"]` | `nil` — **false positive**, "abc" is not fully matched |
| `(re-matches #"(a)" "ab")` | `["a" "a"]` | `nil` — **false positive** |
| `(re-matches #"\d+" "abc")` | `nil` | `nil` |

So it is not merely conservative: a group-less pattern matches only when the
haystack happens to be exactly one byte long, and a pattern whose group count
coincides with the haystack length matches even when the match is a strict
prefix. `re-find`/`re-seq` are unaffected — `(re-find #"\d+" "42")` → `"42"`,
`(re-seq #"\d+" "a1 b22 c333")` → `("1" "22" "333")`.

## Secondary defect in the same branch

When the guard fails, `MatcherState` is left in `MatchPhase::New` rather than
advanced to `Complete`, so `next()` returns `New` forever and the matcher has no
terminal state. `builtin_re_matches` calls `next()` exactly once, so nothing
spins today, but a caller draining a `match_all` matcher until `Complete` would
loop indefinitely. Any fix should set `Complete` on the non-matching path, the
way the `None` arm already does.

## Fix

Java's `Matcher.matches()` — what Clojure's `re-matches` delegates to — requires
the entire region to match. The check wants both ends anchored:

```rust
if !self.match_all || (cap.start() == 0 && cap.end() == self.haystack.get().len())
```

`captures()` returns the leftmost match, so `start() == 0` is needed as well as
`end()`; group 0 can begin part-way into the haystack. After #330 the wrapper in
`crates/cljrs-value/src/regex.rs` exposes `Captures::end()` but not `start()`, so
this is one accessor plus the guard, in one file, and works under either regex
engine.

## Why it survived

`re-matches` appears nowhere in `crates/**/tests`, `tests/`, or
`clojure-test-suite/` — only in its registration (`builtins.rs`) and its
docstring. Worth landing coverage for the anchoring cases in the table above
alongside the fix, including the two false-positive shapes, since a
`start`-only or `end`-only fix still passes the obvious cases.

## Provenance

Pre-existing; not introduced by #330. That PR rewrote the surrounding code to
put an engine-agnostic wrapper behind `Value::Pattern` and deliberately preserved
this line verbatim (renaming `cap.len()` to the clearer
`cap.group_count()`, with a `TODO` recording the mismatch) so the engine swap
stayed behaviour-neutral. On that branch the guard is at
`crates/cljrs-value/src/regex.rs:227`.
